### PR TITLE
Implement the core part of the new-style differentiable function

### PR DESCRIPTION
**Pull request by beam2d**
This PR is the first step of #2871. It implements `FunctionNode` that supports differentiable and economical backprop. The following points are changed.

- `FunctionNode` is added.
- The computational graph in Chainer now consists of `VariableNode` and `FunctionNode`.
- Each variable holds its gradient *variable* instead of gradient *array*. It is stored at `grad_var` attribute as @muupan suggested (the name is slightly simplified). `Variable.grad` is now a proxy to extract the data array of `grad_var`.
  - Note that the owner of the gradient is moved from `VariableNode` to `Variable` so that we can avoid circular references in the computational graph.
- It also adds `FunctionAdapter`, which is an adapter class to convert an old-style function to a new-style one. `Function` is automatically wrapped by `FunctionAdapter` at the forward computation.
- `Variable.backward` is rewritten to support the new-style function interface.
- This PR also supports the new-style function interface for a tiny set tof functions (namely, `+`, `*`, and `F.identity`) for the debugging purpose. It also includes a small test of differentiable backprop using them (now we can differentiate `x * x * x` two times to get `6 * x`). See [here](https://github.com/beam2d/chainer/blob/71e758aa7c874e2ca1aaf59164b6fc3c87274904/tests/chainer_tests/test_variable.py#L201) for the test code.

### CODE REVIEW

**Comment by muupan** [comment]
How gradient penalty (http://arxiv.org/abs/1706.10295) is implemented with this new API?

**Comment by beam2d** [comment]
In the current PR, the code would look like follows.
```py3
x_hat = ...  # interpolation between real and fake data
D = discriminator(x_hat)

creator = x_hat.creator_node
x_hat.creator_node = None  # cut the graph
D.backward()
x_hat.creator_node = creator  # recover the graph

gp = F.sum((F.batch_l2_norm_squared(x_hat.grad_var) ** 0.5 - 1) ** 2)
discriminator.cleargrads()  # NOTE
(gan_loss + _lambda * gp).backward()
```
It's not efficient enough because this PR does not include an API to compute the gradient w.r.t. a specified set of variables (like `grad` of PyTorch). I will work on that in the next step.

NOTE: the `cleargrads` is actually not needed in the current PR because the implementation in this PR does not accumulate the gradient to existing one. It is not compatible with the current Chainer and so I might have to fix it.

**Comment by beam2d** [comment]
(Also note that almost all functions used in the above code do not support differentiable backprop yet, so we also have to work on that)

**Comment by muupan** [comment]
Thank you! I just wanted to confirm the API works for my problems, so it's ok that not all the functions are supported yet.

Let's say `link` is parameterized by `W` and we want to compute ∇_W L(x), where

L(x) = loss(link(x; W - α∇_W loss(link(x; W)))).

That is, L(x) is the loss after an SGD update. Will code like this work?

```
y1 = link(x)
f(y1).backward()
link2 = copy.deepcopy(link)
link2.W = link.W - alpha * link.W.grad_var
y2 = link2(x)
link.cleargrads()  # not needed?
f(y2).backward()
print(link.W)  # ∇_W L(x)
```

**Comment by beam2d** [comment]
I think your code should work.

**Comment by muupan** [comment]
Thank you! Looking forward to it being merged.

File: chainer/links/connection/tree_lstm.py [hunk 1]
```diff
@@ -71,7 +71,7 @@ def __call__(self, *cshsx):
                 :math:`(c_{new}, h_{new})`, where :math:`c_{new}` represents
                 new cell state vector, and :math:`h_{new}` is new output
                 vector.
-
+n
```
Comment (okuta) [inline chainer/links/connection/tree_lstm.py]: Is this typo?

File: tests/chainer_tests/function_hooks_tests/test_debug_print.py [hunk 2]
```diff
@@ -104,6 +104,7 @@ def test_backward_cpu(self):
 - grad: mean=[0-9.\-e]+, std=[0-9.\-e]+$
 '''
         actual = self.io.getvalue()
+        print(actual)
```
Comment (okuta) [inline tests/chainer_tests/function_hooks_tests/test_debug_print.py]: Please remove this line.

**Comment by beam2d** [comment]
I have rebased this PR onto #3019. Now it passes the TimerHook tests.

**Comment by beam2d** [comment]
Rebased onto the current master.

File: chainer/function_node.py [hunk 3]
```diff
@@ -0,0 +1,588 @@
+import collections
+import traceback
+import weakref
+
+import chainer
+from chainer import configuration
+from chainer import cuda
+from chainer import function
```
Comment (beam2d) [inline chainer/function_node.py]: Will do

File: chainer/function_node.py [hunk 4]
```diff
@@ -0,0 +1,588 @@
...
+        not have to use this property; use :meth:`get_retained_outputs`
+        instead.
+
+        """
+        if self._retained_output_data is None:
+            raise RuntimeError(
+                'retained output data has been gone maybe because\n'
+                'One possible cause is that retain_outputs() is called with '
+                'retain_after_backward=False.')
+        out_data = [None for _ in self.outputs]
```
Comment (beam2d) [inline chainer/function_node.py]: It seems that runs faster, so I will use that. Thanks.

File: chainer/function_node.py [hunk 5]
```diff
@@ -0,0 +1,588 @@
...
+            ``len(target_input_indexes)`` or ``len(grad_outputs)``.
+
+        .. seealso::
+
+           :meth:`backward_accumulate` provides an alternative interface that
+           allows you to implement the backward computation fused with the
+           gradient accumulation.
+
+        """
+        return tuple(None for _ in target_input_indexes)
```
Comment (beam2d) [inline chainer/function_node.py]: Will use that

File: chainer/function_node.py [hunk 6]
```diff
@@ -0,0 +1,590 @@
...
+           This behavior might be changed in a future version.
+
+        """
+        # The default implementation uses backward(). You can override this
+        # method without using backward().
+        gxs = self.backward(target_input_indexes, grad_outputs)
+
+        len_gxs = len(gxs)
+        if len_gxs == len(self.inputs):
+            gxs = tuple(gxs[i] for i in target_input_indexes)
```
Comment (unnonouno) [inline chainer/function_node.py]: tuple([])?

File: chainer/function_node.py [hunk 7]
```diff
@@ -0,0 +1,590 @@
...
+        elif len_gxs != len(target_input_indexes):
+            raise ValueError(
+                'number of gradients returned by %s (%s) is incorrect.'
+                % (self._impl_name, self.label))
+
+        return tuple(gx if g_input is None else
```
Comment (unnonouno) [inline chainer/function_node.py]: tuple([])?

File: chainer/function_node.py [hunk 8]
```diff
@@ -0,0 +1,590 @@
...
+                     g_input if gx is None else
+                     gx + g_input
+                     for gx, g_input in zip(gxs, grad_inputs))
```
Comment (unnonouno) [inline chainer/function_node.py]: six

File: chainer/function_node.py [hunk 9]
```diff
@@ -0,0 +1,590 @@
...
+
+        This method is used to retrieve the input variables retained in
+        :meth:`forward`.
+
+        Returns:
+            A tuple of retained input variables.
+
+        """
+        inputs = self.inputs
+        return tuple(inputs[index].get_variable()
```
Comment (unnonouno) [inline chainer/function_node.py]: tuple([])?

File: chainer/function_node.py [hunk 10]
```diff
@@ -0,0 +1,590 @@
...
+           this method creates a fresh variable node that acts as an output
+           node of the function node.
+
+        """
+        ret = []
+        outputs = self.outputs
+
+        new_outputs = list(outputs)
+        outputs_modified = False
+        for index, data in zip(self._output_indexes_to_retain,
```
Comment (unnonouno) [inline chainer/function_node.py]: six

File: chainer/variable.py [hunk 11]
```diff
@@ -675,97 +821,135 @@ def backward(self, retain_grad=False):
...
+            # accumulated to the current gradient w.r.t. the same variable.
+            # In this case, the current implementation passes the current
+            # gradient only to the first occurrence of the variable in the
+            # input tuple and passes ``None`` to the rest of the occurrences.
+            # For example, when the input variables are ``(x, x)``, the
+            # input gradient passed to the ``backward_accumulate`` method is
+            # ``(gx, None)`` where ``gx`` is the current gradient of ``x``.
+            # See also the docstring of ``FunctionNode.backward_accumulate``.
+            target_input_indexes = [
+                i for (i, x) in enumerate(inputs) if x.requires_grad
```
Comment (unnonouno) [inline chainer/variable.py]: ```for i, x in``` ?

File: chainer/function_node.py [hunk 12]
```diff
@@ -0,0 +1,590 @@
...
+       style differentiable functions.
+
+    Attributes:
+        inputs: A tuple of the input :class:`VariableNode` objects.
+        outputs: A tuple of weak references to the output
+            :class:`VariableNode` objects.
+        rank (int): An ordinal following the topological order of the
+            computational graph.
+        stack: Stack trace retrieved at the forward computation. The stack
+            trace is available only in the debug mode.
```
Comment (beam2d) [inline chainer/function_node.py]: I think there is no good pointer to debug mode in the reference now. Is it enough to just add a link to configuration.rst?

File: chainer/function_node.py [hunk 13]
```diff
@@ -0,0 +1,590 @@
...
+        """A tuple of the retained output arrays.
+
+        This property is mainly used by :class:`Function`. Users basically do
+        not have to use this property; use :meth:`get_retained_outputs`
+        instead.
+
+        """
+        if self._retained_output_data is None:
+            raise RuntimeError(
+                'retained output data has been gone maybe because\n'
```
Comment (beam2d) [inline chainer/function_node.py]: It seems this error message is obsolete. I will fix the entire message.

File: chainer/function.py [hunk 14]
```diff
@@ -55,111 +52,126 @@ def force_backprop_mode():
     return configuration.using_config('enable_backprop', True)
 
 
-class Function(object):
-
-    """Function on variables with backpropagation ability.
+class FunctionAdapter(function_node.FunctionNode):
```
Comment (beam2d) [inline chainer/function.py]: Added

File: chainer/function_node.py (hunk 12 above)
Comment (unnonouno) [inline chainer/function_node.py]: Oh, if so it is OK to fix this docstring after we make a document about debug mode.

File: chainer/function_node.py [hunk 15]
```diff
@@ -0,0 +1,591 @@
...
+        .. seealso::
+
+           :meth:`backward_accumulate` provides an alternative interface that
+           allows you to implement the backward computation fused with the
+           gradient accumulation.
+
+        """
+        return (None,) * len(target_input_indexes)
+
+    def backward_accumulate(self, target_input_indexes, grad_outputs,
```
Comment (unnonouno) [inline chainer/function_node.py]: Is it intended to support accumulation of sparse representation of gradients such as EmbedID?

File: chainer/function.py [hunk 16]
```diff
@@ -55,111 +55,129 @@ def force_backprop_mode():
...
+        for x, gx in six.moves.zip(self.inputs, gxs):
+            variable._check_grad_type(self, x, gx)
 
-    For functions that do not need a part of inputs in backward computation,
-    there is a way to possibly reduce the memory consumption by quickly
-    releasing the input arrays after the forward propagation. This is done by
-    calling :meth:`retain_inputs` from inside of :meth:`forward` (including
-    :meth:`forward_cpu` and :meth:`forward_gpu`). See the documentation of
-    :meth:`retain_inputs` for details.
+        return tuple([
```
Comment (unnonouno) [inline chainer/function.py]: In my understand when a user uses old-style Function, this gradient Variable does not have creator link. Is it right?

File: chainer/function_node.py [hunk 17]
```diff
@@ -0,0 +1,591 @@
...
+        for index, data in six.moves.zip(self._output_indexes_to_retain,
+                                         self._retained_output_data):
+            out_data[index] = data
+        return tuple(out_data)
+
+    @property
+    def _impl_name(self):
+        return self.__class__.__name__
+
+    def apply(self, inputs):
```
Comment (beam2d) [inline chainer/function_node.py]: I thought that passing a sequence directly is more verbose and sometimes makes the code simpler (esp. in the case the caller also builds the arguments as a tuple or a list). It is also slightly faster than `*inputs` (25% faster when `inputs` is a tuple, and 100% faster when `inputs` is a list in my env). Is it confusing to use a calling convention different from `Function`?

File: chainer/function_node.py [hunk 18]
```diff
@@ -0,0 +1,591 @@
...
+        forward method.**
+
+        Args:
+            indexes (iterable of int): Indexes of input variables that the
+                function does not require for backprop.
+
+        """
+        self._output_indexes_to_retain = indexes
+
+    def backward(self, target_input_indexes, grad_outputs):
```
Comment (beam2d) [inline chainer/function_node.py]: Hmm, do you come up with a better name? If we rename it, I think we should rename `forward` as well for consistency (I mean the names of `forward` and `backward` should have a clear correspondence).

File: chainer/function_node.py (hunk 15 above)
Comment (beam2d) [inline chainer/function_node.py]: Ah, yes. I didn't originally think of EmbedID, but this interface is basically intended to be used for the cases that accumulation can be made efficient by fusing the operations with backward itself. EmbedID should be one of such cases.

File: chainer/function.py (hunk 16 above)
Comment (beam2d) [inline chainer/function.py]: Right.

**Comment by beam2d** [comment]
I rebased it onto the current master, also squashing the commits.

**Comment by beam2d** [comment]
I added an attribute to VariableNode with which we can automatically detect if a user tried to differentiate an old-style function twice (which should be incorrect). It now correctly raises an error in such a case.

**Comment by unnonouno** [comment]
OK, thank you. That sounds good.

File: chainer/function.py [hunk 19]
```diff
@@ -55,111 +55,138 @@ def force_backprop_mode():
...
-    calling :meth:`retain_inputs` from inside of :meth:`forward` (including
-    :meth:`forward_cpu` and :meth:`forward_gpu`). See the documentation of
-    :meth:`retain_inputs` for details.
 
-    For functions that need a part of outputs in backward computation, it is
-    **strongly recommended** to call :meth:`retain_outputs` from inside of
-    :meth:`forward` (including :meth:`forward_cpu` and :meth:`forward_gpu`).
-    It marks the specified output variable nodes to retain the data. The
-    retained data can be accessed by :attr:`output_arrays` property.
+class Function(object):
```
Comment (delta2323) [inline chainer/function.py]: Are you thinking of removing Function class in the future major version? Although it is not recommended to use this API, its removal seems to affect virtually all existing user code.

Comment (beam2d) [inline chainer/function.py]: At least currently, I do not plan to remove Function.

Comment (beam2d) [inline chainer/function.py]: And, using Function is not "not recommended" at least for now. Function is easier to implement than FunctionNode when one does not need differentiable/economical backward.

Comment (delta2323) [inline chainer/function.py]: Got it. Thanks.
